i18n cleanup 5

mhevery opened the pull request:
This PR builds upon https://github.com/angular/angular/pull/39003. Only review the last `refactor(core):` changes.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
fix(core): Store ICU state in `LView` rather than in `TView`

Before this refactoring/fix the ICU would store the current selected
index in `TView`. This is incorrect, since if ICU is in `ngFor` it will
cause issues in some circumstances. This refactoring properly moves the
state to `LView`.

closes #37021
closes #38144
closes #38073

mhevery committed:
refactor(core): Change `TemplateFixture` to named parameters

`TemplateFixture` used to have positional parameters and many tests got
hard to read as number of parameters reach 10+ with many of them `null`.
This refactoring changes `TemplateFixture` to take named parameters
which improves usability and readability in tests.

mhevery committed:
refactor(core): move `i18n_spec.ts` into `i18n` subfolder

`i18n_spec.ts` file was incorrectly in the `render3` folder rather than `render3/i18n`

mhevery committed:
refactor(core): Change `TName.tagName` to a more generic `value` name.

This is a pre-requisite for making the `TNode.value` a generic storage
mechanism for attaching data to `TNode`.

mhevery committed:
refactor(core): Remove hack where we `TIcu` was stored in `tagName`

Remove casting where we stored `TIcu` in `TNode.tagName` which was of
type `string` rather than `TIcu'. (renamed to `TNode.value` in previous
commit.)

mhevery committed:
refactor(core): Create `TNodeType.Text` to display full template in `TView` debug

When looking at `TView` debug template only Element nodes were displayed
as `TNode.Element` was used for both `RElement` and `RText`.
Additionally no text was stored in `TNode.value`. The result was that
the whole template could not be reconstructed. This refactoring creates
`TNodeType.Text` and store the text value in `TNode.value`.

The refactoring also changes `TNodeType` into flag-like structure make
it more efficient to check many different types at once.

mhevery committed:
refactor(core): Rename `debugMatch` to `matchDebug` for consistency

Previous function name `debugMatch` was not consistent with other match
functions.

alfaproject commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -177,9 +177,10 @@ function createTNodeAndAddOpCode(
     ensureIcuContainerVisitorLoaded(loadIcuContainerVisitor);
   }
-  createOpCodes.push(opCode, text);
+  createOpCodes.push(opCode, text === null ? '' : text);
```
```suggestion
  createOpCodes.push(opCode, text || '');
```

or `??`

mhevery replied:
my understanding is that the first form is more performant because it is checking if text is `null`. The second form is slower because it is checking for truthiness, which is more complicated. For example empty string.

Will leave as is.

alfaproject commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -177,9 +177,10 @@ function createTNodeAndAddOpCode(
+  createOpCodes.push(opCode, text === null ? '' : text);
   const tNode = createTNodeAtIndex(
-      tView, i18nNodeIdx, isICU ? TNodeType.IcuContainer : TNodeType.Element, null, null);
+      tView, i18nNodeIdx, isICU ? TNodeType.Icu : TNodeType.Text, text === null ? '{{?}}' : text,
```
```suggestion
      tView, i18nNodeIdx, isICU ? TNodeType.Icu : TNodeType.Text, text || '{{?}}',
```

mhevery replied:
same here.

alfaproject commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -52,21 +69,26 @@ export const enum TNodeType {
+  Placeholder = 0b1000000,
+
+  // Combined Types
+  AnyRNode = 0b11,        // Text | Element,
+  AnyContainer = 0b1100,  // Container | ElementContainer,
```
```suggestion
  AnyRNode = TNodeType.Text | TNodeType.Element,
  AnyContainer = TNodeType.Container | TNodeType.ElmentContainer,
```

mhevery replied:
Yes I tried that, but unfortunately it does not work. I get this error:
```
error TS2535: Enum type 'TNodeType' has members with initializers that are not literals.
```

mhevery replied:
added.

alfaproject commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -508,9 +509,7 @@ export function getClosestRElement(tView: TView, tNode: TNode|null, lView: LView
-  while (parentTNode != null &&
-         (parentTNode.type === TNodeType.ElementContainer ||
-          parentTNode.type === TNodeType.IcuContainer)) {
+  while (parentTNode != null && (parentTNode.type & (TNodeType.ElementContainer | TNodeType.Icu))) {
```
I know it was already there but is the `!=` on purpose?

mhevery replied:
Nope, fixed it.

mhevery committed:
refactor(core): rename `COMMENT_MARKER` to `ICU_MARKER`

`COMMENT_MARKER` is a generic name which does not make it obvious that
it is used for ICU use case. `ICU_MARKER` is more explicit as it is used
exclusively with ICUs.

mhevery committed:
refactor(core): cleanup i18n/icu data structures

- Made `*OpCodes` array branded for safer type checking.
- Simplify `I18NRemoveOpCodes` encoding.
- Broke out  `IcuCreateOpCodes` from `I18nMutableOpCodes`.

mhevery committed:
refactor(core): Consistent use of  `HEADER_OFFSET` (in `ɵɵ*` instructions only)

IMPORTANT: `HEADER_OFFSET` should only be refereed to the in the `ɵɵ*` instructions to translate
instruction index into `LView` index. All other indexes should be in the `LView` index space and
there should be no need to refer to `HEADER_OFFSET` anywhere else.

mhevery committed:
refactor(core): Use `~x` instead of `-x` which can result in `-0`

`expandoInstructions` uses negative numbers by `-x`. This has lead to
issues in the paste as `-0` is processed as float rather than integer
leading to de-optimization.

AndrewKushnir commented on packages/core/src/render3/util/view_utils.ts:
```diff
@@ -119,14 +121,14 @@ export function getNativeByTNodeOrNull(tNode: TNode|null, lView: LView): RNode|n
   ngDevMode && assertGreaterThan(index, -1, 'wrong index for TNode');
-  ngDevMode && assertLessThan(index, tView.data.length - HEADER_OFFSET, 'wrong index for TNode');
-  return tView.data[index + HEADER_OFFSET] as TNode;
+  ngDevMode && assertLessThan(index, tView.data.length, 'wrong index for TNode');
```
It looks like we can use the same checks like we have in `getNativeByIndex`?
```suggestion
  ngDevMode && assertIndexInRange(tView.data, index);
  ngDevMode && assertGreaterThanOrEqual(index, HEADER_OFFSET, 'Expected to be past HEADER_OFFSET');
```

mhevery replied:
👍

AndrewKushnir commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -390,8 +390,8 @@ export function createContainerRef(
-        const commentTNode = createTNode(
-            hostView[TVIEW], hostTNode.parent, TNodeType.Container, hostTNode.type, null, null);
+        const commentTNode =
+            createTNode(hostView[TVIEW], hostTNode.parent, TNodeType.Container, 0, null, null);
```
Would it make sense to use `-1` (instead of `0`) here as an indication that this is just a dummy index?

mhevery replied:
I tried that, but unfortunately no. The issue is that `0` is a proper read into `LView`/`TData` array but `-1` is not, which results in `undefined` and than a lot of code fails. It we want to use `-1` I would have to add lots of checks which I think are unnecessary.

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -710,16 +708,17 @@ function walkIcuTree(
   if (depth === 0) {
-    remove.push(index << I18nMutateOpCode.SHIFT_REF | I18nMutateOpCode.RemoveNestedIcu);
-    remove.push(index << I18nMutateOpCode.SHIFT_REF | I18nMutateOpCode.Remove);
+    remove.push(~index);
+    remove.push(index);
```
Would be great to add a quick comment here.

mhevery replied:
done.

AndrewKushnir commented on packages/core/src/render3/instructions/i18n_icu_container_visitor.ts:
```diff
@@ -63,16 +63,13 @@ export function loadIcuContainerVisitor() {
-            assertEqual(opCode, I18nMutateOpCode.RemoveNestedIcu, 'Expecting RemoveNestedIcu');
         _stack.push(_index, _removes);
-        const tIcu = _lView[TVIEW].data[removeOpCode >>> I18nMutateOpCode.SHIFT_REF] as TIcu;
+        const tIcu = _lView[TVIEW].data[~removeOpCode] as TIcu;
```
```suggestion
        // ICUs are represented by negative indices
        const tIcuIndex = ~removeOpCode;
        const tIcu = _lView[TVIEW].data[tIcuIndex] as TIcu;
```

mhevery replied:
done.

AndrewKushnir commented on packages/core/src/render3/instructions/i18n_icu_container_visitor.ts:
```diff
@@ -54,7 +54,7 @@ export function loadIcuContainerVisitor() {
       _removes = tIcu.remove[currentCase];
     } else {
-      _removes = EMPTY_ARRAY;
+      _removes = EMPTY_ARRAY as any;
```
```suggestion
      _removes = EMPTY_ARRAY as I18nRemoveOpCodes;
```

mhevery replied:
Sorry same problem, can't do that.

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_debug.ts:
```diff
@@ -21,8 +21,8 @@ import {ELEMENT_MARKER, getInstructionFromI18nMutateOpCode, getParentFromI18nMut
-    this: I18nUpdateOpCodes|void, opcodes?: I18nUpdateOpCodes): string[] {
-  const createOpCodes: I18nUpdateOpCodes = opcodes || (Array.isArray(this) ? this : []);
+    this: I18nCreateOpCodes|void, opcodes?: I18nCreateOpCodes): string[] {
+  const createOpCodes: I18nCreateOpCodes = opcodes || (Array.isArray(this) ? this : [] as any);
```
```suggestion
  const createOpCodes: I18nCreateOpCodes = opcodes || (Array.isArray(this) ? this : [] as I18nCreateOpCodes);
```

mhevery replied:
Yes, that does not work. You first have to cast to `x as any as I18nCreateOpCodes` for it to work. That is why I have `any` on right` but `I18nCreateOpCodes` on left.

AndrewKushnir commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -69,8 +69,8 @@ export function i18nStartFirstCreatePass(
   const rootTNode = getCurrentParentTNode();
-  const createOpCodes: I18nCreateOpCodes = [];
-  const updateOpCodes: I18nUpdateOpCodes = [];
+  const createOpCodes: I18nCreateOpCodes = [] as any;
```
Can we avoid `as any` here (and in other similar cases)?
```suggestion
  const createOpCodes = [] as I18nCreateOpCodes;
```

mhevery replied:
unfortunately no. you can't cast `[]` to `I18nCreateOpCodes` directly without going through `any`.

petebacondarwin reviewed (commented):
Typo in **refactor(core): Use `~x` instead of `-x` which can result in `-0`**

`paste` -> `past`

petebacondarwin commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -47,7 +47,13 @@ export const DECLARATION_COMPONENT_VIEW = 16;
+/**
+ * Size of LView's header. Necessary to adjust for it when setting slots.
+ *
+ * IMPORTANT: `HEADER_OFFSET` should only be refereed to the in the `ɵɵ*` instructions to translate
```
```suggestion
 * IMPORTANT: `HEADER_OFFSET` should only be referred to the in the `ɵɵ*` instructions to translate
```

mhevery replied:
👍

petebacondarwin commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -604,12 +604,12 @@ export function i18nParseTextIntoPartsAndICU(pattern: string): (string|IcuExpres
-  const create: I18nMutateOpCodes = [];
-  const remove: I18nMutateOpCodes = [];
-  const update: I18nUpdateOpCodes = [];
+  const create: IcuCreateOpCodes = [] as any;
+  const remove: I18nRemoveOpCodes = [] as any;
+  const update: I18nUpdateOpCodes = [] as any;
```
NIT you can just do:

```suggestion
  const create = [] as IcuCreateOpCodes;
  const remove = [] as I18nRemoveOpCodes;
  const update = [] as I18nUpdateOpCodes;
```

mhevery replied:
nope, that does not work, because of the brand.

petebacondarwin replied:
OH! Sorry. I didn't realise.

petebacondarwin commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -428,7 +428,7 @@ export interface TNode {
 
   /** The tag name associated with this node. */
-  value: string|null;
+  value: any;
```
Did you consider making this a generic type? E.g.

```ts
export interface TNode<TValue = string|null> {
  ...
  value: TValue;
  ...
}
```

And then you can have

```ts
export interface TIcuContainerNode extends TNode<TIcu> { ... }
```

mhevery replied:
The subclasses have proper type.

petebacondarwin replied:
Can you explain this? What subclasses do you mean?

mhevery replied:
`TIcuContainerNode` overrides `value` to be `TIcu`. So result is the same thing without the generics.

mhevery commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -428,9 +428,7 @@ export interface TNode {
   providerIndexes: TNodeProviderIndexes;
 
   /** The tag name associated with this node. */
```
👍

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -25,7 +25,7 @@ import {isFactory, NO_PARENT_INJECTOR, NodeInjectorFactory, NodeInjectorOffset,
 import {isComponentDef, isComponentHost} from './interfaces/type_checks';
 import {DECLARATION_COMPONENT_VIEW, DECLARATION_VIEW, INJECTOR, LView, T_HOST, TData, TVIEW, TView, TViewType} from './interfaces/view';
-import {assertNodeOfPossibleTypes} from './node_assert';
+import {assertNodeType} from './node_assert';
```
👍

mhevery commented on packages/core/src/render3/i18n/i18n_parse.ts:
```diff
@@ -177,9 +177,10 @@ function createTNodeAndAddOpCode(
+  createOpCodes.push(opCode, text === null ? '' : text);
   const tNode = createTNodeAtIndex(
-      tView, i18nNodeIdx, isICU ? TNodeType.IcuContainer : TNodeType.Element, null, null);
+      tView, i18nNodeIdx, isICU ? TNodeType.Icu : TNodeType.Text, text === null ? '{{?}}' : text,
```
Good idea! done.

mhevery commented on packages/core/src/render3/instructions/i18n.ts:
```diff
@@ -65,7 +65,7 @@ export function ɵɵi18nStart(
   // If `parentTNode` is an `ElementContainer` than it has `<!--ng-container--->`.
   // When we do inserts we have to make sure to insert in front of `<!--ng-container--->`.
-  const insertInFrontOf = parentTNode && parentTNode.type === TNodeType.ElementContainer ?
+  const insertInFrontOf = parentTNode && parentTNode.type & TNodeType.ElementContainer ?
```
👍

mhevery commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -223,7 +223,7 @@ export function getOrCreateTNode(
+    tView: TView, index: number, type: TNodeType.Icu, name: null,
     attrs: TAttributes|null): TElementContainerNode;
 export function getOrCreateTNode(
     tView: TView, index: number, type: TNodeType, name: string|null, attrs: TAttributes|null):
```
good idea! done.

mhevery replied:
done

mhevery commented on packages/core/src/render3/instructions/template.ts:
```diff
@@ -69,6 +69,8 @@ export function ɵɵtemplate(
     tagName?: string|null, attrsIndex?: number|null, localRefsIndex?: number|null,
     localRefExtractor?: LocalRefExtractor) {
+  // FIXME(misko): tagName is meaningless because the compiler never generates it. It should be
+  // removed.
```
thanks, remove comment.

mhevery commented on packages/core/src/render3/query.ts:
```diff
@@ -217,7 +217,7 @@ class TQuery_ implements TQuery {
       // to go past `<ng-container>` to determine <i #target> parent node (but we shouldn't traverse
       // up past the query's host node!).
-      while (parent !== null && parent.type === TNodeType.ElementContainer &&
+      while (parent !== null && parent.type & TNodeType.ElementContainer &&
```
👍

mhevery commented on packages/core/src/render3/query.ts:
```diff
@@ -253,7 +253,7 @@ class TQuery_ implements TQuery {
       if (read !== null) {
         if (read === ViewEngine_ElementRef || read === ViewContainerRef ||
-            read === ViewEngine_TemplateRef && tNode.type === TNodeType.Container) {
+            read === ViewEngine_TemplateRef && tNode.type & TNodeType.Container) {
```
👍

mhevery commented on packages/core/src/render3/i18n/i18n_apply.ts:
```diff
@@ -364,41 +364,31 @@ function applyIcuSwitchCase(tView: TView, tIcu: TIcu, lView: LView, value: strin
+        ngDevMode && assertDefined(rNode, 'RNode expected here');
+        ngDevMode && assertDomNode(rNode);
+        nativeRemoveNode(lView[RENDERER], rNode);
+        ngDevMode && ngDevMode.rendererRemoveNode++;
```
done

mhevery commented on packages/core/src/render3/i18n/i18n_debug.ts:
```diff
@@ -100,35 +100,31 @@ export function i18nUpdateOpCodesToString(
+ * @param this `I18nCreateOpCodes` if attached as a method.
+ * @param opcodes `I18nCreateOpCodes` if invoked as a function.
  */
 export function i18nMutateOpCodesToString(
```
good catch, done.

mhevery commented on packages/core/src/render3/component.ts:
```diff
@@ -172,9 +171,10 @@ export function createRootComponentView(
+  const index = HEADER_OFFSET;
+  ngDevMode && assertIndexInRange(rootView, index);
+  rootView[index] = rNode;
+  const tNode: TElementNode = getOrCreateTNode(tView, index, TNodeType.Element, '#host', null);
```
Why? I like seeing it in debug, it makes it clear that it is not a regular node.

AndrewKushnir replied:
Oh, got it, thanks Misko. I thought it was left there after some debugging. May be we can add a quick comment here to mention that it's a debug-only string?

mhevery commented on packages/core/src/render3/component.ts:
```diff
@@ -233,7 +233,7 @@ export function createRootComponent<T>(
   if (tView.firstCreatePass &&
       (componentDef.hostBindings !== null || componentDef.hostAttrs !== null)) {
-    const elementIndex = rootTNode.index - HEADER_OFFSET;
+    const elementIndex = rootTNode.index;
     setSelectedIndex(elementIndex);
```
done

mhevery commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -228,11 +228,12 @@ export function getOrCreateTNode(
 export function getOrCreateTNode(
     tView: TView, index: number, type: TNodeType, name: string|null, attrs: TAttributes|null):
     TElementNode&TContainerNode&TElementContainerNode&TProjectionNode&TIcuContainerNode {
+  ngDevMode && index !== 0 &&  // 0 are bogus nodes and they are OK.
```
done

mhevery commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -885,7 +887,7 @@ export function createTNode(
       {
         type: type,
-        index: adjustedIndex,
+        index: index,
```
done.

mhevery commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1731,6 +1732,7 @@ function markTransplantedViewsForRefresh(lView: LView) {
  * @param componentHostIdx  Element index in LView[] (adjusted for HEADER_OFFSET)
  */
 function refreshComponent(hostLView: LView, componentHostIdx: number): void {
+  ngDevMode && assertNotEqual(hostLView[0], hostLView[1] as any, '????');
```
Opps, I added that while debugging, should be reverted.

AndrewKushnir reviewed (approved):
LGTM, thanks Misko! 👍 (just one minor comment)

mhevery commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -52,22 +69,28 @@ export const enum TNodeType {
+  Placeholder = 0b1000000,
+
+  // Combined Types
+  AnyRNode = 0b11,        // Text | Element,
```
added.

IgorMinar commented on goldens/circular-deps/packages.json:
```diff
@@ -968,25 +971,37 @@
+    "packages/core/src/render3/interfaces/i18n.ts",
+    "packages/core/src/render3/interfaces/view.ts",
+    "packages/core/src/render3/interfaces/query.ts",
+    "packages/core/src/render3/interfaces/node.ts"
```
I thought that the circular deps increase was just temporary and would go away in the final version of this PR. Is there a way to decrease these circular deps?

mhevery replied:
let me look into it.

mhevery replied:
Fixed, added another commit to this PR which fixes it.

IgorMinar reviewed (approved):
Many of these changes look like a great cleanup! Thanks @mhevery for improving the code base.

I've reviewed the PR at a high level, and spot checked particular areas in detail. I don't have enough prior knowledge of the internals of ICU to provide you with a detailed review in a timely manner, so I'll rely on the other reviewers primarily @AndrewKushnir and @petebacondarwin for doing an in depth review.

The only major concern I have is that I don't see any tests with ng-for that would prove that the original bug has been resolved. Can you please consider adding them?

The size increase is now down to ~1KB and I understand that there is an additional PR that will decrease the size further? if that's the case then I'm fine with the regression.

Lastly, I see a lot of focus on runtime performance but I don't see any benchmarks. Have you measured the runtime perf of the non-i18n and i18n use-cases to ensure that we are not regressing? 

I'm going to give this a LGTM so that we can land this monster PR, but let's follow up on any of the unresolved issues. Also, let's keep the target as `minor` so that it is not included into stable version until 11.1.

Thank you!

Reviewed-for: fw-core, size-tracking

AndrewKushnir commented:
Hi @IgorMinar,

> Lastly, I see a lot of focus on runtime performance but I don't see any benchmarks. Have you measured the runtime perf of the non-i18n and i18n use-cases to ensure that we are not regressing?

Great question, thanks for mentioning that! Yes, we did come up with i18n scenarios for perf tests (added as a [separate PR here](https://github.com/angular/angular/pull/39142)). We also ran all other benchmarks and there were no performance regressions.

IgorMinar commented:
Awesome! Thank you @AndrewKushnir!

mhevery commented on packages/core/src/render3/node_assert.ts:
```diff
@@ -6,44 +6,40 @@
 
-export function assertNodeOfPossibleTypes(
-    tNode: TNode|null, types: TNodeType[], message?: string): void {
+export function assertNotTNodeType(
```
deleted

mhevery committed:
refactor(core): clean up circular dependencies

Moved code from `interfaces/i18n.ts` which was causing circular dependencies

mhevery commented:
> The only major concern I have is that I don't see any tests with ng-for that would prove that the original bug has been resolved. Can you please consider adding them?

Good catch! Hmm, that is so weird. I had a test for that which I copied from repro. Not sure how it got lost. re-added it.

petebacondarwin reviewed (approved):
Reviewed-for: global-docs-approvers

mhevery commented:
[presubmit](http://test/OCL:337689497:BASE:337689503:1602976626024:c8ed3e27)

jelbourn reviewed (approved):
LGTM exclusively for size-tracking

Reviewed-for: size-tracking

mhevery commented:
[presubmit](http://test/OCL:338273380:BASE:338306634:1603305645977:c0fec186)

mhevery commented:
[presubmit](http://test/OCL:338273380:BASE:338317716:1603308830999:2cc128bd)

mhevery commented:
Closing in favor of https://github.com/angular/angular/pull/39301

mhevery commented:
[presubmit](http://test/OCL:338273380:BASE:338360285:1603321826652:dac938ed)

alxhub reviewed (approved):
Reviewed-for: public-api

jelbourn reviewed (approved):
LGTM exclusively for the public-api change

Reviewed-for: public-api

AndrewKushnir reviewed (approved):
Reviewed-for: public-api
